README
Tree::Simple::View version 0.16
===========================

See the individual module documentation for more information

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

	Tree::Simple

COPYRIGHT AND LICENCE

Copyright (C) 2004-2007 Infinity Interactive, Inc.

http://www.iinteractive.com

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

Changes
Revision history for Perl extension Tree-Simple-View.

0.17 
    * added node_formatter support for ASCII view

0.16 Mon. July, 9, 2007
    * cleaning bad files in the tar.gz
    * converted to Build.PL

0.15 Mon. May 21, 2007
    * added a simple Tree::Simple::View::ASCII
        - added docs and tests for it

0.14 Mon Novermber 28, 2005
    - removing the OS X resource fork files
    - applied patch (RT #16064) to support XHTML 
      compliant output for Tree::Simple::View::HTML.
        - thanks to Simon Wilcox for this patch and tests

0.13 Mon. November 7, 2005
    - fixing typo on the generated HTML (rt.cpan.org Ticket #14963)
    - cleaned up View.pm as little as well

0.12 Mon Jan 31 2005
    - converted this to use Class::Throwable for exceptions
       - altered tests for this
       
    - added the 'radio_button' and 'checkbox' attributes to
      the Tree::Simple::View::DHTML class
        - added tests for this
        - added documentation for this

0.11 Thurs Nov 18 2004
    - now using Scalar::Util::blessed() instead of 
      the convoluted UNIVERSAL::isa() stuff. 
      - added Scalar::Util as a dependency

0.10 Sat Aug 7 2004
    - Added the setPathComparisonFunction method
      so that custom node value's can be compared
      easily. Thanks to Neyuki for this idea.
        - documentented and tested this

0.09 Mon Aug 2 2004
    - I forgot to change the version requirements for 
      Tree::Simple to 1.07 which is the first version 
      that uses the getUID methods.

0.08 Sun Aug 1 2004
    - Changed the way that the DHTML DIV IDs were being
      created, you can now use your own UID from the
      Tree::Simple object instead, thanks to 
      Brett Nuske for that suggestion.
        - documented and tested it

0.07 Thurs July 15 2004
    - added includeTrunk method, and implemented the
      necessary functionality in ::HTML and ::DHTML.
      - added tests to check this new functionality
        as well
      - added documentation on this new functionality

0.06 Fri July 2 2004
    - Added a number of tests to the suite as well as 
      improved some existing tests.

0.05 Fri Jun 25 2004
    - changes the HREF part of the DHTML links to be
      javascript:void(0) rather than just #, it stops
      a problem in some browsers where clicking a link
      would scroll you to the top of the page.
      - updated tests to reflect this change
    - removed some code which was not nessecary from the
      Tree::Simple::View::HTML class. 

0.04 Fri Jun 25 2004
    !*!*!*! MAJOR UPDATE !*!*!*!
    - My orginal vision for Tree::Simple::View was that of
      a Proxy/Factory/I-Dont-Know-What-I-Was-Thinking class.
      But the idea really wasn't that good, matter of fact
      it was pretty bad. So this version marks a major update
      to the module, the details of which are below.
      - Tree::Simple::View::Base has been removed, its code
        has been moved to Tree::Simple::View.
      - Tree::Simple::View is no longer a "Proxy" class (not
        that it ever really was) for the Tree::Simple::View::*
        classes. It now is the abstract base class, taking on
        the role that Tree::Simple::View::Base once served.
      - Tree::Simple::View::HTML has been updated to reflect 
        its new base class. (Tree::Simple::View::DHTML is a 
        subclass of Tree::Simple::View::HTML, so it didn't need
        to be changed). 
      - All documentation has been updated to reflect these changes.
    - I have added a bunch of tests while I was at it.

0.03 Wed Jun 23 2004
    - fixed a number of little issues with the DHTML class
        - fixed problems with combined CSS configurations
          and the expand-collapse functionality
        - fixed a bug whereby expandPath did not work
          properly
    - added an example file of the DHTML output 
      (examples/DHTML_example.html)
    - added some documentation here and there

0.02 Tues Jun 22 2004
    - added simple accessors for the tree and configurations
      to the Base class
    - added the 'form_element_formatter' to the DHTML class
      which allows for form elements to be added to the tree
      outside of the link tag. 
    - fixed a bug in the DHTML implementation where the DHTML
      would not work if you had not set a link_css value. Now
      you it does work :)

0.01 Fri Jun 18 14:38:19 2004
    - module created



lib/Tree/Simple/View/ASCII.pm

package Tree::Simple::View::ASCII;

use strict;
use warnings;

our $VERSION = '0.02';

use base 'Tree::Simple::View';

use Tree::Simple::View::Exceptions;

sub expandPathSimple  {
    my ($self, $tree, @full_path) = @_;
        
    my $output = '';    
    my @vert_dashes;

    my $traversal = sub {
        my ($t, $redo, $current_path, @path) = @_;
        $output .= $self->_processNode($t, \@vert_dashes)
            unless $t->isRoot; 
        foreach my $child ($t->getAllChildren()) {     
            if (defined $current_path && $self->_compareNodeToPath($current_path, $child)) {
                $output .= $redo->($child, $redo, @path);
            }
            else {
                $output .= $self->_processNode($child, \@vert_dashes); 
            }
        }
    };
    
    $output .= $self->_processNode($tree, \@vert_dashes)
        if $self->{include_trunk};    
    
    if ($self->{include_trunk} && defined $full_path[0] && $self->_compareNodeToPath($full_path[0], $tree)) {
        shift @full_path;
    }
    
    # Its the U combinator baby!
    $traversal->($tree, $traversal, @full_path);        

    return $output;
}

sub expandAllSimple  {
    my ($self) = @_;   
    
    my $output = '';    
    my @vert_dashes;

    $output .= $self->_processNode($self->{tree}, \@vert_dashes)
        if $self->{include_trunk};

    $self->{tree}->traverse(sub {
        my $t = shift;
        $output .= $self->_processNode($t, \@vert_dashes);
    });
    
    return $output;
}

sub expandPathComplex { 
    my ($self, $tree, undef, @full_path) = @_;
    # complex stuff is not supported here ...
    $self->expandPathSimple($tree, @full_path);
}

*expandAllComplex  = \&expandAllSimple;

sub _processNode {
    my ($self, $t, $vert_dashes) = @_;
    
    my $depth         = $t->getDepth;
    my $sibling_count = $t->isRoot ? 1 : $t->getParent->getChildCount;    
    
    $depth++ if $self->{include_trunk};
    
    my @indent = map {
        $vert_dashes->[$_] || "        "
    } 0 .. $depth - 1;

    @$vert_dashes = (
        @indent, 
        ($sibling_count == 1 
            ? ("        ") 
            : ("    |   "))
    );

    if ($sibling_count == ($t->getIndex + 1)) {
        $vert_dashes->[$depth] = "        ";
    }

    my $node = exists $self->{config}->{node_formatter} 
        ? $self->{config}->{node_formatter}->($t)
        : $t->getNodeValue;

    return ((join "" => @indent[1 .. $#indent]) 
            . ($depth ? "    |---" : "") 
            . $node 
            . "\n");    
}

1;

__END__

=pod

=head1 NAME

Tree::Simple::View::ASCII - A class for viewing Tree::Simple hierarchies in ASCII

=head1 SYNOPSIS

  use Tree::Simple::View::ASCII;
  
  my $tree_view = Tree::Simple::View::ASCII->new($tree);
  
  $tree_view->includeTrunk(1);
  
  print $tree_view->expandAll();
  
  # root
  #     |---Node Level: 1
  #     |       |---Node Level: 1.1
  #     |       |---Node Level: 1.2
  #     |       |       |---Node Level: 1.2.1
  #     |       |       |---Node Level: 1.2.2
  #     |       |---Node Level: 1.3
  #     |---Node Level: 2
  #     |       |---Node Level: 2.1
  #     |       |---Node Level: 2.2
  #     |---Node Level: 3
  #     |       |---Node Level: 3.1
  #     |               |---Node Level: 3.1.1
  #     |               |---Node Level: 3.1.2
  #     |---Node Level: 4
  #             |---Node Level: 4.1

  print $tree_view->expandPath("root", "Node Level: 1", "Node Level: 1.2");

  # root
  #     |---Node Level: 1
  #     |       |---Node Level: 1.1
  #     |       |---Node Level: 1.2
  #     |       |       |---Node Level: 1.2.1
  #     |       |       |---Node Level: 1.2.2
  #     |       |---Node Level: 1.3
  #     |---Node Level: 2
  #     |---Node Level: 3
  #     |---Node Level: 4

=head1 DESCRIPTION

This is a Tree::Simple::View subclass which provides simple ASCII formatting for trees.

I had this code lying around, and I figured it was best to put it into here. This is 
an early release of this, and it lacks configuration parameter handling, but that can 
be easily added later when I need it.

=head1 METHODS

=over 4

=item B<expandAllSimple>

This will draw a fully expanded tree. 

=item B<expandAllComplex>

This currently is aliased to the C<expandAllSimple> since we don't have config handling.

=item B<expandPathSimple (@path)>

This will draw a tree with the given C<@path> expanded. 

=item B<expandPathComplex>

This currently is aliased to the C<expandPathSimple> since we don't have config handling.

=back

=head1 TODO

Config handling, allowing you to customize the drawing. Patches welcome, I just don't 
currently have the time to add it.

=head1 BUGS

None that I am aware of. Of course, if you find a bug, let me know, and I will be sure to fix it. 

=head1 AUTHOR

stevan little, E<lt>stevan@iinteractive.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2008 by Infinity Interactive, Inc.

L<http://www.iinteractive.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut


lib/Tree/Simple/View/Exceptions.pm

package Tree::Simple::View::Exceptions;

use strict;
use warnings;

our $VERSION = '0.01';

use Class::Throwable qw(
    Tree::Simple::View::InsufficientArguments
    Tree::Simple::View::AbstractMethod
    Tree::Simple::View::AbstractClass
    Tree::Simple::View::CompilationFailed
);
    
1;

__END__

=pod

=head1 NAME

Tree::Simple::View::Exceptions - A set of exceptions for Tree::Simple::View

=head1 SYNOPSIS

  use Tree::Simple::View::Exceptions;

=head1 DESCRIPTION

This just creates and loads a few exceptions for use by the Tree::Simple::View classes. 
Nothing else to see really.

=head1 EXCEPTIONS

=over 4

=item B<Tree::Simple::View::InsufficientArguments>

=item B<Tree::Simple::View::AbstractMethod>

=item B<Tree::Simple::View::AbstractClass>

=item B<Tree::Simple::View::CompilationFailed>

=back

=head1 BUGS

None that I am aware of. Of course, if you find a bug, let me know, and I will be sure to fix it. 

=head1 CODE COVERAGE

See the CODE COVERAGE section of Tree::Simple::View for details.

=head1 SEE ALSO

=over 4

=item L<Class::Throwable>

=back

=head1 AUTHOR

stevan little, E<lt>stevan@iinteractive.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2008 by Infinity Interactive, Inc.

L<http://www.iinteractive.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut


lib/Tree/Simple/View/HTML.pm

package Tree::Simple::View::HTML;

use strict;
use warnings;

our $VERSION = '0.12';

use base 'Tree::Simple::View';

use Tree::Simple::View::Exceptions;

use constant OPEN_TAG  => 1;
use constant CLOSE_TAG => 2;
use constant EXPANDED  => 3;

my %tags = (
    xhtml => { OL => 'ol', UL => 'ul', LI => 'li', STYLE => q{ style='}, CLASS => q{ class='} },
    html  => { OL => 'OL', UL => 'UL', LI => 'LI', STYLE => q{ STYLE='}, CLASS => q{ CLASS='} },
);

## public methods

sub expandPathSimple  { 
    my ($self, $tree, $current_path, @path) = @_;
    my @results;
    # if we were not called from this routine, and 
    # include trunk has been turned on then, this is 
    # the first time we have been called, so ...
    if ($self->{include_trunk} && (caller(1))[3] !~ /expandPathSimple$/) {    
        push @results => "<UL>";
        push @results => ("<LI>" . $tree->getNodeValue() . "</LI>");     
        # now recurse but dont change any of the args,
        # (if we are supposed to that is, based on the path)   
        push @results => ($self->expandPathSimple($tree, @path))
            if (defined $current_path && $tree->getNodeValue() eq $current_path); 
        push @results => "</UL>";               
    }
    else {
        push @results => "<UL>";
        foreach my $child ($tree->getAllChildren()) {
            if (defined $current_path && $self->_compareNodeToPath($current_path, $child)) {
                push @results => ("<LI>" . $child->getNodeValue() . "</LI>");
                push @results => ($self->expandPathSimple($child, @path));
            }
            else {
                push @results => ("<LI>" . $child->getNodeValue() . "</LI>");
            }
        }
        push @results => "</UL>";
    }
    return (join "\n" => @results);        
}

sub expandPathComplex {
    my ($self, $tree, $config, $current_path, @path) = @_;
    # get the config
    my ($list_func, $list_item_func) = $self->_processConfig($config);  
    
    # use the helper function to recurse
    my $_expandPathComplex = sub {
        my ($self_func, $list_func, $list_item_func, $tree, $current_path, @path) = @_;
        my @results = ($list_func->(OPEN_TAG));
        foreach my $child ($tree->getAllChildren()) {
            if (defined $current_path && $self->_compareNodeToPath($current_path, $child)) {
                unless ($child->isLeaf()) {
                    push @results => ($list_item_func->($child, EXPANDED));
                    push @results => ($self_func->($self_func, $list_func, $list_item_func, $child, @path));
                }
                else {
                    push @results => ($list_item_func->($child));
                }
            }
            else {
                push @results => ($list_item_func->($child));
            }
        }
        push @results => ($list_func->(CLOSE_TAG));
        return (join "\n" => @results);   
    };
    
    my @results;
    if ($self->{include_trunk}) {    
        push @results => ($list_func->(OPEN_TAG));
        if (defined $current_path && $self->_compareNodeToPath($current_path, $tree)) {      
            push @results => ($list_item_func->($tree, EXPANDED));     
            push @results => $_expandPathComplex->($_expandPathComplex, $list_func, $list_item_func, $tree, @path);
        }
        else {
           push @results => ($list_item_func->($tree));     
        }
        push @results => ($list_func->(CLOSE_TAG));             
    } 
    else {
        push @results => $_expandPathComplex->($_expandPathComplex, $list_func, $list_item_func, $tree, $current_path, @path);            
    }    
    
    return (join "\n" => @results);   
}

sub expandAllSimple  {
    my ($self) = @_;   
    my @results = ("<UL>");
    my $root_depth = $self->{tree}->getDepth() + 1;    
    my $last_depth = -1;
    my $traversal_sub = sub {
        my ($t) = @_;
        my $current_depth = $t->getDepth();
        push @results => ("</UL>" x ($last_depth - $current_depth)) if ($last_depth > $current_depth);
        push @results => ("<LI>" . $t->getNodeValue() . "</LI>");
        push @results => "<UL>" unless $t->isLeaf();
        $last_depth = $current_depth;
    };
    $traversal_sub->($self->{tree}) if $self->{include_trunk};
    $self->{tree}->traverse($traversal_sub);
    $last_depth -= $root_depth;    
    $last_depth++ if $self->{include_trunk}; 
    push @results => ("</UL>" x ($last_depth + 1)); 
    return (join "\n" => @results);
}

sub expandAllComplex {
    my ($self, $config) = @_;
    
    my ($list_func, $list_item_func) = $self->_processConfig($config);   
    
    my @results = $list_func->(OPEN_TAG);
    my $root_depth = $self->{tree}->getDepth() + 1;    
    my $last_depth = -1;
    my $traversal_sub = sub {
        my ($t) = @_;
        my $current_depth = $t->getDepth();
        push @results => ($list_func->(CLOSE_TAG) x ($last_depth - $current_depth)) if ($last_depth > $current_depth);
        if ($t->isLeaf()) {
            push @results => ($list_item_func->($t));
        }
        else {
            push @results => ($list_item_func->($t, EXPANDED));
        }
        push @results => $list_func->(OPEN_TAG) unless $t->isLeaf();
        $last_depth = $current_depth;
    };
    $traversal_sub->($self->{tree}) if $self->{include_trunk};
    $self->{tree}->traverse($traversal_sub);
    $last_depth -= $root_depth;    
    $last_depth++ if $self->{include_trunk};  
    push @results => ($list_func->(CLOSE_TAG) x ($last_depth + 1)); 
    return (join "\n" => @results);
}

## private methods

# process configurations

sub _processConfig {
    my ($self, $config) = @_;
    my %config = %{$config};
    
    # Make sure the tag style is always set to something we know &
    # set tags to be the hashref of tags we want to save extra indirection later
    if ( !exists $config{ tag_style } ) {
        $config{ tags } = $tags{ html };
    }
    elsif ( !exists( $tags{ $config{ tag_style } }) ) {
        throw Tree::Simple::View::CompilationFailed "unknown tag_style $config{ tag_style }";
    }
    else {
        $config{ tags } = $tags{ $config{ tag_style } };
    }
        
    my $list_func = $self->_buildListFunction(%config) 
        || throw Tree::Simple::View::CompilationFailed "list function didn't compile", $@;
    my $list_item_func  = $self->_buildListItemFunction(%config) 
    	|| throw Tree::Simple::View::CompilationFailed "list item function didn't compile", $@;

    return ($list_func, $list_item_func);
}

## code strings to be evaluated

use constant LIST_FUNCTION_CODE_STRING => q|
    sub {
        my ($tag_type) = @_;
        return '<' . $config{tags}->{$list_type} . ${list_css} . '>' if ($tag_type == OPEN_TAG);
        return '</' . $config{tags}->{$list_type} .'>' if ($tag_type == CLOSE_TAG);
    }
|;

use constant LIST_ITEM_FUNCTION_CODE_STRING  => q|;
    sub {
        my ($t, $is_expanded) = @_;
        my $item_css = $list_item_css;
        $item_css = $expanded_item_css if ($is_expanded && $expanded_item_css);
        return '<'.$config{tags}->{LI}.$item_css.'>' . (($node_formatter) ? $node_formatter->($t) : $t->getNodeValue()) . '</'.$config{tags}->{LI}.'>';
    }
|;

## list config processing
sub _processListConfig {
    my ($self, %config) = @_;
    
    my $list_type = "UL";
    $list_type = (($config{list_type} eq "unordered") ? "UL" : "OL") if exists $config{list_type};

    my $list_css = "";
    if (exists $config{list_css}) {
        # make sure we have a proper ';' at the end
        # of the CSS code here, it is needed by the
        # DHTML subclass when we add the display property
        # to it, no other element requires this so far,
        # but if it did, this same idiom could be reused
        my $_list_css = $config{list_css};
        $_list_css .= ";" unless ($_list_css =~ /\;$/);        
        $list_css = $config{tags}->{STYLE} . "${_list_css}'";
    }
    elsif (exists $config{list_css_class}) {
        $list_css = $config{tags}->{CLASS} . $config{list_css_class} . "'";
    }
    # otherwise do nothing and stick with default
    
    return ($list_type, $list_css);
}

sub _buildListFunction {
    my ($self, %config) = @_;
    # process the configuration directives
    my ($list_type, $list_css) = $self->_processListConfig(%config);
    # now compile the subroutine in the current environment
    return eval $self->LIST_FUNCTION_CODE_STRING;
}

## list item config processing

sub _processListItemConfig {
    my ($self, %config) = @_;
    
    my $list_item_css = "";
    if (exists $config{list_item_css}) {
        $list_item_css = $config{tags}->{STYLE} . $config{list_item_css} . "'";
    }
    elsif (exists $config{list_item_css_class}) {
        $list_item_css = $config{tags}->{CLASS} . $config{list_item_css_class} . "'";
    }
    # otherwise do nothing and stick with default    

    my $expanded_item_css = "";
    if (exists $config{expanded_item_css}) {
        $expanded_item_css = $config{tags}->{STYLE} . $config{expanded_item_css} . "'";
    }
    elsif (exists $config{expanded_item_css_class}) {
        $expanded_item_css = $config{tags}->{CLASS} . $config{expanded_item_css_class} . "'";
    }
    # otherwise do nothing and stick with default    
    
    my $node_formatter;
    $node_formatter = $config{node_formatter} 
        if (exists $config{node_formatter} && ref($config{node_formatter}) eq "CODE");
        
    return ($list_item_css, $expanded_item_css, $node_formatter);
}

sub _buildListItemFunction {
    my ($self, %config) = @_;
    # process the configuration directives
    my ($list_item_css, $expanded_item_css, $node_formatter) = $self->_processListItemConfig(%config);
    # now compile the subroutine in the current environment    
    return eval $self->LIST_ITEM_FUNCTION_CODE_STRING;
}

1;

__END__

=pod

=head1 NAME

Tree::Simple::View::HTML - A class for viewing Tree::Simple hierarchies in HTML

=head1 SYNOPSIS

  use Tree::Simple::View::HTML;
  
  ## a simple example
  # use the defaults (an unordered list with no CSS)
  my $tree_view = Tree::Simple::View::HTML->new($tree);

  ## more complex examples
                                        
  # use the CSS properties                                    
  my $tree_view = Tree::Simple::View::HTML->new($tree => (
                                list_type  => "ordered",
                                list_css => "list-style: circle;",
                                list_item_css => "font-family: courier;",
                                expanded_item_css => "font-family: courier; font-weight: bold",                               
                                ));  
                                
  # use the CSS classes                                  
  my $tree_view = Tree::Simple::View::HTML->new($tree => (
                                list_css_class => "myListClass",
                                list_item_css_class => "myListItemClass",
                                expanded_item_css_class => "myExpandedListItemClass",                                
                                ));                                                                                                                                                                                                                                                                                                                                                                                                                                                                                
                                   
  # mix the CSS properties and CSS classes
  my $tree_view = Tree::Simple::View::HTML->new($tree => (
                                list_css => "list-style: circle;",
                                list_item_css => "font-family: courier;",
                                expanded_item_css_class => "myExpandedListItemClass",                                                         
                                node_formatter => sub {
                                    my ($tree) = @_;
                                    return "<B>" . $tree->getNodeValue()->description() . "</B>";
                                    }
                                ));  
                              
  
  # print out the tree fully expanded
  print $tree_view->expandAll();
  
  # print out the tree expanded along a given path (see below for details)
  print $tree_view->expandPath("Root", "Child", "GrandChild");                                                           

=head1 DESCRIPTION

This is a class for use with Tree::Simple object hierarchies to serve as a means of 
displaying them in HTML. It is the "View", while the Tree::Simple object hierarchy 
would be the "Model" in your standard Model-View-Controller paradigm. 

This class outputs fairly vanilla HTML in its simpliest configuration, suitable for 
both legacy browsers and text-based browsers. Through the use of various configuration 
options, CSS can be applied to support more advanced browsers but still degrade 
gracefully to legacy browsers. 

=head1 METHODS

=over 4

=item B<new ($tree, %configuration)>

Accepts a C<$tree> argument of a Tree::Simple object (or one derived from Tree::Simple), 
if C<$tree> is not a Tree::Simple object, and exception is thrown. This C<$tree> object 
does not need to be a ROOT, you can start at any level of the tree you desire. The 
options in the C<%config> argument are as follows:

=over 4

=item I<tag_style>

This can be either 'html' or 'xhtml', which will produce output with tags in capitals 
or lowercase respectively, for xhtml compliance. The default is 'html' for backwards 
compatibility.

=item I<list_type>

This can be either 'ordered' or 'unordered', which will produce ordered and unordered 
lists respectively. The default is 'unordered'.

=item I<list_css>

This can be a string of CSS to be applied to the list tag (C<UL> or C<OL> depending 
upon the I<list_type> option). This option and the I<list_css_class> are mutually 
exclusive, and this option will override in a conflict.

=item I<list_css_class>

This can be a CSS class name which is applied to the list tag (C<UL> or C<OL> depending 
upon the I<list_type> option). This option and the I<list_css> are mutually exclusive, 
and the I<list_css> option will override in a conflict.

=item I<list_item_css>

This can be a string of CSS to be applied to the list item tag (C<LI>). This option 
and the I<list_item_css_class> are mutually exclusive, and this option will override 
in a conflict.

=item I<list_item_css_class>

This can be a CSS class name which is applied to the list item tag (C<LI>). This option 
and the I<list_item_css> are mutually exclusive, and the I<list_item_css> option will 
override in a conflict.

=item I<expanded_item_css>

This can be a string of CSS to be applied to the list item tag (C<LI>) if it has an 
expanded set of children. This option and the I<expanded_item_css_class> are mutually 
exclusive, and this option will override in a conflict.

=item I<expanded_item_css_class>

This can be a CSS class name which is applied to the list item tag (C<LI>) if it has 
an expanded set of children. This option and the I<expanded_item_css> are mutually 
exclusive, and the I<expanded_item_css> option will override in a conflict.

=item I<node_formatter>

This can be a CODE reference which will be given the current tree object as its only 
argument. The output of this subroutine will be placed within the list item tags 
(C<LI>). This option can be used to implement; custom formatting of the node, handling 
of complex node objects or implementing any type of handler code to drive your 
interface (using link tags or form submissions, etc). 

=back

=item B<getTree>

A basic accessor to reach the underlying tree object. 

=item B<getConfig>

A basic accessor to reach the underlying configuration hash. 

=item B<includeTrunk ($boolean)>

This controls the getting and setting (through the optional C<$boolean> argument) of 
the option to include the tree's trunk in the output. Many times, the trunk is not 
actually part of the tree, but simply a root from which all the branches spring. 
However, on occasion, it might be nessecary to view a sub-tree, in which case, the 
trunk is likely intended to be part of the output. This option defaults to off. 

=item B<setPathComparisonFunction ($CODE)>

This takes a C<$CODE> reference, which can be used to add custom path comparison 
features to Tree::Simple::View. The function will get two arguments, the first is 
the C<$current_path>, the second is the C<$current_tree>. When using C<expandPath>, 
it may sometimes be nessecary to be able to control the comparison of the path values. 
For instance, your node may be an object and need a specific method called to match 
the path against. 

=item B<expandPath (@path)>

This method will return a string of HTML which will represent your tree expanded 
along the given C<@path>. This is best shown visually. Given this tree:

  Tree-Simple-View
      lib
          Tree
              Simple
                  View.pm
                  View
                      HTML.pm
                      DHTML.pm
      Makefile.PL
      MANIFEST
      README 
      Changes
      t
          10_Tree_Simple_View_test.t
          20_Tree_Simple_View_HTML_test.t
          30_Tree_Simple_View_DHTML_test.t
          
And given this path:

  Tree-Simple-View, lib, Tree, Simple

Your display would like something like this:

  Tree-Simple-View
      lib
          Tree
              Simple
                  View.pm
                  View
      Makefile.PL
      MANIFEST
      README 
      Changes
      t

As you can see, the given path has been expanded, but no other sub-trees are 
shown (nor is the HTML of the un-expanded nodes to be found in the output). 

It should be noted that this method actually calls either the C<expandPathSimple> 
or C<expandPathComplex> method depending upon the C<%config> argument in the 
constructor. See their documenation for details.

=item B<expandPathSimple ($tree, @path)>

If no C<%config> argument is given in the constructor, then this method is called 
by C<expandPath>. This method is optimized since it does not need to process any 
configuration, but just as the name implies, it's output is simple.

This method can also be used for another purpose, which is to bypass a previously 
specified configuration and use the base "simple" configuration instead.

=item B<expandPathComplex ($tree, $config, @path)>

If a C<%config> argument is given in the constructor, then this method is called 
by C<expandPath>. This method has been optimized to be used with configurations, 
and will actually custom compile code (using C<eval>) to speed up the generation 
of the output.

This method can also be used for another purpose, which is to bypass a previously 
specified configuration and use the configuration specified (as a HASH reference) 
in the C<$config> parameter.

=item B<expandAll>

This method will return a string of HTML which will represent your tree completely 
expanded.

It should be noted that this method actually calls either the C<expandAllSimple> 
or C<expandAllComplex> method depending upon the C<%config> argument in the 
constructor.   

=item B<expandAllSimple>

If no C<%config> argument is given in the constructor, then this method is called 
by C<expandAll>. This method too is optimized since it does not need to process 
any configuration.

This method as well can also be used to bypass a previously specified configuration 
and use the base "simple" configuration instead.

=item B<expandAllComplex ($config)>

If a C<%config> argument is given in the constructor, then this method is called 
by C<expandAll>. This method too has been optimized to be used with configurations, 
and will also custom compile code (using C<eval>) to speed up the generation of 
the output.

Just as with C<expandPathComplex>, this method can be to bypass a previously 
specified configuration and use the configuration specified (as a HASH reference) 
in the C<$config> parameter.

=back

=head1 TO DO

=over 4

=item B<depth-based css>

I would like to be able to set any of my css properties as an array, which would 
essentially allow for depth-based css values. For instance, something like this:

  list_css => [
      "font-size: 14pt;",
      "font-size: 12pt;",
      "font-size: 10pt;"      
      ];

This would result in the first level of the tree having a font-size of 14 points, 
the second level would have a font-size of 12 points, then all other levels past 
the second level (third and beyond) would have a font-size of 10 points. Of course 
if a fourth element were added to this array (ex: "font-size: 8pt;"), then the third 
level would have a font-size of 10 points, and all others past that level would 
have the font-size of 8 points. 

Ideally this option would be available for all I<*_css> and I<*_css_class> options. 
I have not yet figured out the best way to do this though, so ideas/suggestions are 
welcome, of course, patches are even better.

=back

=head1 BUGS

None that I am aware of. Of course, if you find a bug, let me know, and I will be 
sure to fix it. 

=head1 CODE COVERAGE

See the CODE COVERAGE section of Tree::Simple::View for details.

=head1 SEE ALSO

If a DHTML based tree is what you are after, then look at the Tree::Simple::View::DHTML class.

A great CSS reference can be found at:

    http://www.htmlhelp.com/reference/css/

Information specifically about CSS for HTML lists is at:

    http://www.htmlhelp.com/reference/css/classification/list-style.html

=head1 AUTHOR

stevan little, E<lt>stevan@iinteractive.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2008 by Infinity Interactive, Inc.

L<http://www.iinteractive.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut
